rust/agents/relayer/src/msg/gas_payment/mod.rs

```diff
@@ -18,72 +21,112 @@ mod policies;
-            GasPaymentEnforcementPolicy::Minimum { payment } => {
-                Box::new(GasPaymentPolicyMinimum::new(payment))
-            }
-            GasPaymentEnforcementPolicy::MeetsEstimatedCost { coingeckoapikey } => {
-                Box::new(GasPaymentPolicyMeetsEstimatedCost::new(coingeckoapikey))
-            }
-        };
+        let policies = policy_configs
+            .into_iter()
+            .map(|cfg| {
+                let p: Box<dyn GasPaymentPolicy> = match cfg.policy {
+                    GasPaymentEnforcementPolicy::None => Box::new(GasPaymentPolicyNone::new()),
+                    GasPaymentEnforcementPolicy::Minimum { payment } => {
+                        Box::new(GasPaymentPolicyMinimum::new(payment))
+                    }
+                    GasPaymentEnforcementPolicy::MeetsEstimatedCost { coingeckoapikey } => {
+                        Box::new(GasPaymentPolicyMeetsEstimatedCost::new(coingeckoapikey))
+                    }
+                    GasPaymentEnforcementPolicy::OnChainFeeQuoting { gasfraction } => {
+                        let gasfraction = gasfraction.replace(' ', "");
+                        let v: Vec<&str> = gasfraction.split('/').collect();
+                        assert_eq!(
+                            v.len(),
+                            2,
+                            r#"Could not parse gas fraction; expected "`numerator / denominator`""#
+                        );
+                        Box::new(GasPaymentPolicyOnChainFeeQuoting::new(
+                            v[0].parse::<u64>().expect("Invalid integer"),
+                            v[1].parse::<u64>().expect("Invalid integer"),
+                        ))
+                    }
+                };
+                (p, cfg.whitelist)
+            })
+            .collect();
 
-        Self {
-            policy,
-            whitelist,
-            db,
-        }
+        Self { policies, db }
     }
 }
 
 impl GasPaymentEnforcer {
-    /// Returns (gas payment requirement met, current payment according to the DB)
+    /// Returns Some(gas_limit) if the enforcer has approved the transaction or
+    /// None if the transaction is not approved.
     pub async fn message_meets_gas_payment_requirement(
         &self,
         message: &HyperlaneMessage,
         tx_cost_estimate: &TxCostEstimate,
-    ) -> Result<(bool, U256)> {
-        let current_payment = self.get_message_gas_payment(message.id())?;
+    ) -> Result<Option<U256>> {
+        let msg_id = message.id();
+        let current_payment = self.db.retrieve_gas_payment_for_message_id(msg_id)?;
+        let current_expenditure = self.db.retrieve_gas_expenditure_for_message_id(msg_id)?;
 
-        // If the message matches the whitelist, consider it as meeting the gas payment requirement
-        if self.whitelist.msg_matches(message, false) {
-            return Ok((true, current_payment));
+        for (policy, whitelist) in &self.policies {
```

Comment:
may be nice to have some debug / trace logs here that could help figure out if messages aren't meeting gas payment requirements